language/stdlib/modules/transaction_fee.move

```diff
@@ -1,105 +1,87 @@
+        Transaction::assert(Signer::address_of(fee_account) == 0xFEE, 0);
+        move_to(fee_account, LBRIdent<LBR>{ coin: Libra::zero<LBR>() });
     }
 
-    // After the book keeping has been performed, this then distributes the
-    // transaction fees equally to all validators with the exception that
-    // any remainder (in the case that the number of validators does not
-    // evenly divide the transaction fee pot) is distributed to the first
-    // validator.
-    fun distribute_transaction_fees_internal<Token>(
-        amount_to_distribute_per_validator: u64,
-        num_validators: u64,
-    ) acquires TransactionFees {
-        let distribution_resource = borrow_global<TransactionFees>(0xFEE);
-        let index = 0;
+    // Called for LibraAccount setup
+    public fun grant_fee_payment_capability(libra_account_singleton_addr: &signer): FeePaymentCapability {
+        Transaction::assert(Signer::address_of(libra_account_singleton_addr) == 0xA550C18, 0);
+        FeePaymentCapability {}
+    }
 
-        while (index < num_validators) {
+    public fun pay_fee<CoinType>(fee: Libra<CoinType>, _cap: &FeePaymentCapability)
+    acquires CollectedFees, LBRIdent {
+        // For LBR we need to extract the underlying coins from the
+        // on-chain reserve and then push these to the
+        if (exists<LBRIdent<CoinType>>(0xFEE)) {
 
-            let addr = LibraSystem::get_ith_validator_address(index);
-            // Increment the index into the validator set.
-            index = index + 1;
+            // The cointype is LBR, but we need to cast CoinType to LBR in
+            // order to extract the underlying coins. We do this by going
+            // through storage.
+            let lbr_ident = borrow_global_mut<LBRIdent<CoinType>>(0xFEE);
+            let value = Libra::value(&fee);
+            Libra::deposit(&mut lbr_ident.coin, fee);
+            let lbr_fee = Libra::withdraw<LBR>(
+                &mut borrow_global_mut<LBRIdent<LBR>>(0xFEE).coin,
+                value,
+            );
 
-            LibraAccount::pay_from_capability<Token>(
-                addr,
-                &distribution_resource.fee_withdrawal_capability,
-                amount_to_distribute_per_validator,
-                x"",
-                x""
+            let (coin1, coin2) = LBR::unpack(lbr_fee);
+            let collected_fee_coin1 = borrow_global_mut<CollectedFees<Coin1>>(0xFEE);
+            Libra::preburn_with_resource(
+                coin1,
+                &mut collected_fee_coin1.preburn,
+                0xFEE
             );
-           }
+            let collected_fee_coin2 = borrow_global_mut<CollectedFees<Coin2>>(0xFEE);
+            Libra::preburn_with_resource(
+                coin2,
+                &mut collected_fee_coin2.preburn,
+                0xFEE
+            );
+        } else {
+            let collected_fee = borrow_global_mut<CollectedFees<CoinType>>(0xFEE);
+            Libra::preburn_with_resource(
```

Comment:
Just realized: I think this will add a new preburn request to the `Preburn.requests` for every transaction. That will get big fast, and it will also require one `BurnTxnFees` transaction for reach fee-paying transaction. We probably want to design this differently, but not sure what the best approach is:
1. Pay fees into an DD-like Association account, which can then preburn/burn to liquidate the fees
2. Don't pay fees into the `Preburn` resource; use a dedicated resource + add logic for burning from that instead
3. Something else?

My bias is toward (1) because it seems simplest. It does mean that the Association could choose to send the funds somewhere instead of burning them, but that seems ok.

Reply:
You are so right... this will add a preburn request for each txn...which is definitely not good.

What about this (which I think is 1?):
* `0xFEE` holds coins (or some resource containing coins) for each currency.
* Transaction fees are deposited into these.
* the `0xB1E55ED` account would hold withdrawal capability on `0xFEE`. Since it has a`BurnCapability` for each cointype (except LBR, and we can do the unpacking to handle this)
* The `0xB1E55ED` account would have preburn resources for each currency != LBR.

We can then have the transaction be something like:
```rust
fun main<CoinType>(blessed: &signer) {
    // uses the withdrawal capability underneath
	let coins = TransactionFee::withdraw_fees<CoinType>();
    Libra::preburn_to_sender(coins);
    Libra::burn(Signer::address_of(blessed));
}
```

Reply:
Love this solution!

Reply:
This should be implemented now. A couple notes:
* The burning is per currency (since that's what sounded would be most in line with AOS). Not hard to change if we decide against this though.
* The transaction fees are stored in the balance's for the account at `0xFEE`.
* the TC account now holds preburn resources for the fiat currencies. We still need to special case LBR in the script, but it's not nearly as awful as before.